runtime/caml/misc.h

```diff
@@ -286,6 +286,19 @@ CAMLnoret CAMLextern void caml_failed_assert (char *, char_os *, int);
 #define CAMLassert(x) ((void) 0)
 #endif
 
+#ifdef CAML_INTERNALS
+#if (defined(__STDC_VERSION__) && __STDC_VERSION__ < 202311L) && \
+      !defined(__cplusplus)
+  #if __has_builtin(__builtin_unreachable)
+    #define unreachable() (__builtin_unreachable())
+  #elif defined(_MSC_VER)
+    #define unreachable() (__assume(0))
+  #else
+    #define unreachable() (abort())
```

Comment:
I'd rather see this defined as
```suggestion
    #define unreachable() (assert(0))
```
so that, should it ever fire, this gets recognized as an assertion failure rather than a mysterious `SIGABRT`.

(And yes, given how eager clang and gcc are at arm-wrestling the C specs in order to win at picobenchmarks, I am expecting such unreachable code paths to be reachable when compiled with these compilers at some of their optimization levels).

Reply:
But what if C assertions are disabled with `NDEBUG`? The code path may then be missing a return statement since there won't be a noreturn function to end it.
Similarly, if we used `CAMLassert` instead, it would result in an empty statement in non-debug runtimes, triggering the same problem.
If I keep an assertion and add a return statement, I'll probably get warnings about an unreachable return statement after the unreachable annotation…

Reply:
There is no simple way to solve this without making every compiler combination happy. You probably should not remove the few `return` following `unreachable` then.

Reply:
The compilers differ too much in behavior… but in my test they complain about missing returns rather that unreachable code. I've changed my code to use `CAMLassert` here.